boa/integrations/jupyter/browser.py

```diff
@@ -0,0 +1,177 @@
+        Then, it waits for the signature to be received via the API.
+        :param tx_data: The transaction data to sign.
+        :return: The signed transaction data.
+        """
+        sign_data = _javascript_call(
+            "signTransaction", tx_data, timeout_message=TRANSACTION_TIMEOUT_MESSAGE
+        )
+        return convert_frontend_dict(sign_data)
+
+
+class BrowserRPC(RPC):
+    """
+    An RPC object that sends requests to the browser via Javascript.
+    """
+
+    @property
+    def identifier(self) -> str:
+        return type(self).__name__  # every instance does the same
+
+    @property
+    def name(self):
+        return self.identifier
+
+    def fetch(self, method: str, params: Any) -> Any:
+        return _javascript_call(
+            "rpc", method, params, timeout_message=RPC_TIMEOUT_MESSAGE
+        )
+
+    def fetch_multi(self, payloads: list[tuple[str, Any]]) -> list[Any]:
+        return _javascript_call(
+            "multiRpc", payloads, timeout_message=RPC_TIMEOUT_MESSAGE
+        )
+
+    def wait_for_tx_receipt(self, tx_hash, timeout: float, poll_latency=1):
+        # we do the polling in the browser to avoid too many callbacks
+        # each callback generates currently 10px empty space in the frontend
+        timeout_ms, pool_latency_ms = timeout * 1000, poll_latency * 1000
+        return _javascript_call(
+            "waitForTransactionReceipt",
+            tx_hash,
+            timeout_ms,
+            pool_latency_ms,
+            timeout_message=RPC_TIMEOUT_MESSAGE,
+        )
+
+
+def _javascript_call(js_func: str, *args, timeout_message: str) -> Any:
+    """
+    This function attempts to call a Javascript function in the browser and then
+    wait for the result to be sent back to the API.
+    - Inside Google Colab, it uses the eval_js function to call the Javascript function.
+    - Outside, it uses a SharedMemory object and polls until the frontend called our API.
+    A custom timeout message is useful for user feedback.
+    :param snippet: A function that given a token and some kwargs, returns a Javascript snippet.
+    :param kwargs: The arguments to pass to the Javascript snippet.
+    :return: The result of the Javascript snippet sent to the API.
+    """
+    install_jupyter_javascript_triggers()
+
+    token = _generate_token()
+    args_str = ", ".join(json.dumps(p) for p in chain([token], args))
+    js_code = f"window._titanoboa.{js_func}({args_str})"
+    # logging.warning(f"Calling {js_func} with {args_str}")
+
+    if colab_eval_js:
+        result = colab_eval_js(js_code)
+        return _parse_js_result(json.loads(result))
+
+    memory = SharedMemory(name=token, create=True, size=SHARED_MEMORY_LENGTH)
+    logging.info(f"Waiting for {token}")
+    try:
+        memory.buf[:1] = NUL
+        display(Javascript(js_code))
+        return _wait_buffer_set(memory.buf, timeout_message)
```

Comment:
it might be nice to have `_wait_buffer_set` return the raw result (without `_parse_js_result`) so it is the responsibility of the caller -- this way both branches in this function all need to call `_parse_js_result`.